src/Sylius/Component/Core/Cart/Provider/NewCartProvider.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\Cart\Provider;
+
+use Sylius\Component\Cart\Model\CartInterface;
+use Sylius\Component\Cart\Provider\CartProviderInterface;
+use Sylius\Component\Channel\Context\ChannelNotFoundException;
+use Sylius\Component\Core\Context\ShopperContextInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+
+/**
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ */
+class NewCartProvider implements CartProviderInterface
```

Comment:
ImmutableCartProvider? Considering [this line](https://github.com/Sylius/Sylius/pull/5197/files#diff-574f7dbf46d15fff5561f80bb9121870R55) we should not name it NewCartProvider. Or CurrentCartProvider? At the end, this class return a current cart provider, if it instance exists, or create a new one.